File: app/src/main/java/org/dhis2/usescases/eventsWithoutRegistration/eventCapture/eventCaptureFragment/EventCaptureFormFragment.java
package org.dhis2.usescases.eventsWithoutRegistration.eventCapture.eventCaptureFragment;

import static org.dhis2.commons.Constants.EVENT_MODE;
import static org.dhis2.commons.extensions.ViewExtensionsKt.closeKeyboard;
import static org.dhis2.usescases.eventsWithoutRegistration.eventCapture.ui.NonEditableReasonBlockKt.showNonEditableReasonMessage;
import static org.dhis2.utils.granularsync.SyncStatusDialogNavigatorKt.OPEN_ERROR_LOCATION;

import android.content.Context;
import android.content.Intent;
import android.os.Bundle;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.databinding.DataBindingUtil;
import androidx.fragment.app.FragmentTransaction;

import org.dhis2.R;
import org.dhis2.commons.Constants;
import org.dhis2.commons.featureconfig.data.FeatureConfigRepository;
import org.dhis2.commons.featureconfig.model.Feature;
import org.dhis2.databinding.SectionSelectorFragmentBinding;
import org.dhis2.form.model.EventMode;
import org.dhis2.form.model.EventRecords;
import org.dhis2.form.ui.FormView;
import org.dhis2.usescases.eventsWithoutRegistration.eventCapture.EventCaptureAction;
import org.dhis2.usescases.eventsWithoutRegistration.eventCapture.EventCaptureActivity;
import org.dhis2.usescases.eventsWithoutRegistration.eventCapture.EventCaptureContract;
import org.dhis2.usescases.general.FragmentGlobalAbstract;
import org.jetbrains.annotations.NotNull;

import javax.inject.Inject;

import kotlin.Unit;

public class EventCaptureFormFragment extends FragmentGlobalAbstract implements EventCaptureFormView,
        OnEditionListener {

    @Inject
    EventCaptureFormPresenter presenter;

    @Inject
    FeatureConfigRepository featureConfig;

    private EventCaptureActivity activity;
    private SectionSelectorFragmentBinding binding;
    private FormView formView;

    public static EventCaptureFormFragment newInstance(
            String eventUid,
            Boolean openErrorSection,
            EventMode eventMode
    ) {
        EventCaptureFormFragment fragment = new EventCaptureFormFragment();
        Bundle args = new Bundle();
        args.putString(Constants.EVENT_UID, eventUid);
        args.putBoolean(OPEN_ERROR_LOCATION, openErrorSection);
        args.putString(EVENT_MODE, eventMode.name());
        fragment.setArguments(args);
        return fragment;
    }

    @Override
    public void onAttach(@NotNull Context context) {
        super.onAttach(context);
        this.activity = (EventCaptureActivity) context;
        activity.eventCaptureComponent.plus(
                new EventCaptureFormModule(
                        this,
                        getArguments().getString(Constants.EVENT_UID))
        ).inject(this);
        setRetainInstance(true);
    }

    @Override
    public void onCreate(@Nullable @org.jetbrains.annotations.Nullable Bundle savedInstanceState) {
        String eventUid = getArguments().getString(Constants.EVENT_UID, "");
        EventMode eventMode = EventMode.valueOf(getArguments().getString(EVENT_MODE));
        formView = new FormView.Builder()
                .locationProvider(locationProvider)
                .onLoadingListener(loading -> {
                    if (loading) {
                        activity.showProgress();
                    } else {
                        activity.hideProgress();
                    }
                    return Unit.INSTANCE;
                })
                .onFocused(() -> {
                    activity.hideNavigationBar();
                    return Unit.INSTANCE;
                })
                .onPercentageUpdate(percentage -> {
                    activity.updatePercentage(percentage);
                    return Unit.INSTANCE;
                })
                .onDataIntegrityResult(result -> {
                    presenter.handleDataIntegrityResult(result);
                    return Unit.INSTANCE;
                })
                .factory(activity.getSupportFragmentManager())
                .setRecords(new EventRecords(eventUid, eventMode))
                .openErrorLocation(getArguments().getBoolean(OPEN_ERROR_LOCATION, false))
                .useComposeForm(
                        featureConfig.isFeatureEnable(Feature.COMPOSE_FORMS)
                )
                .build();
        activity.setFormEditionListener(this);
        super.onCreate(savedInstanceState);
    }

    @Nullable
    @Override
    public View onCreateView(@NonNull LayoutInflater inflater, @Nullable ViewGroup container, @Nullable Bundle savedInstanceState) {
        binding = DataBindingUtil.inflate(inflater, R.layout.section_selector_fragment, container, false);
        EventCaptureContract.Presenter activityPresenter = activity.getPresenter();
        binding.setPresenter(activityPresenter);

        activityPresenter.observeActions().observe(getViewLifecycleOwner(), action ->
        {
            if (action == EventCaptureAction.ON_BACK) {
                formView.onSaveClick();
                activityPresenter.emitAction(EventCaptureAction.NONE);
            }
        });

        binding.actionButton.setOnClickListener(view -> {
            closeKeyboard(view);
            performSaveClick();
        });

        return binding.getRoot();
    }

    @Override
    public void onViewCreated(@NonNull @NotNull View view, @Nullable @org.jetbrains.annotations.Nullable Bundle savedInstanceState) {
        super.onViewCreated(view, savedInstanceState);
        FragmentTransaction transaction = getChildFragmentManager().beginTransaction();
        transaction.replace(R.id.formViewContainer, formView).commit();
        formView.setScrollCallback(isSectionVisible -> {
            animateFabButton(isSectionVisible);
            return Unit.INSTANCE;
        });
    }

    @Override
    public void onResume() {
        super.onResume();
        presenter.showOrHideSaveButton();
    }

    @Override
    public void onActivityResult(int requestCode, int resultCode, @Nullable Intent data) {
        super.onActivityResult(requestCode, resultCode, data);
    }

    private void animateFabButton(boolean sectionIsVisible) {
        int translationX = 1000;
        if (sectionIsVisible) translationX = 0;

        binding.actionButton.animate().translationX(translationX).setDuration(500).start();
    }

    @Override
    public void performSaveClick() {
        formView.onSaveClick();
    }

    @Override
    public void onEditionListener() {
        formView.onEditionFinish();
    }

    @Override
    public void hideSaveButton() {
        binding.actionButton.setVisibility(View.GONE);
    }

    @Override
    public void showSaveButton() {
        binding.actionButton.setVisibility(View.VISIBLE);
    }

    @Override
    public void onReopen() {
        formView.reload();
    }

    @Override
    public void showNonEditableMessage(@NonNull String reason, boolean canBeReOpened) {
        showNonEditableReasonMessage(
                binding.editableReasonContainer,
                reason,
                canBeReOpened,
                () -> {
                    presenter.reOpenEvent();
                    return Unit.INSTANCE;
                }
        );
    }

    @Override
    public void hideNonEditableMessage() {
        binding.editableReasonContainer.removeAllViews();
    }
}

File: app/src/main/java/org/dhis2/usescases/eventsWithoutRegistration/eventCapture/eventCaptureFragment/EventCaptureFormFragment.kt
package org.dhis2.usescases.eventsWithoutRegistration.eventCapture.eventCaptureFragment

import android.content.Context
import android.os.Bundle
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.ui.Modifier
import androidx.databinding.DataBindingUtil
import org.dhis2.R
import org.dhis2.commons.Constants
import org.dhis2.commons.extensions.closeKeyboard
import org.dhis2.databinding.SectionSelectorFragmentBinding
import org.dhis2.form.model.EventMode
import org.dhis2.form.model.EventRecords
import org.dhis2.form.model.RowAction
import org.dhis2.form.ui.FormView
import org.dhis2.mobile.commons.ui.NonEditableReasonBlock
import org.dhis2.usescases.eventsWithoutRegistration.eventCapture.EventCaptureAction
import org.dhis2.usescases.eventsWithoutRegistration.eventCapture.EventCaptureActivity
import org.dhis2.usescases.general.FragmentGlobalAbstract
import org.dhis2.utils.granularsync.OPEN_ERROR_LOCATION
import javax.inject.Inject

class EventCaptureFormFragment :
    FragmentGlobalAbstract(),
    EventCaptureFormView {
    @Inject
    lateinit var presenter: EventCaptureFormPresenter

    private lateinit var activity: EventCaptureActivity

    private lateinit var binding: SectionSelectorFragmentBinding

    private var formView: FormView? = null

    override fun onAttach(context: Context) {
        super.onAttach(context)
        this.activity = context as EventCaptureActivity
        activity.eventCaptureComponent
            ?.plus(
                EventCaptureFormModule(
                    this,
                    arguments?.getString(Constants.EVENT_UID)!!,
                ),
            )?.inject(this)
    }

    override fun onCreate(savedInstanceState: Bundle?) {
        val eventUid = requireArguments().getString(Constants.EVENT_UID, "")
        val eventMode = arguments?.getString(Constants.EVENT_MODE)?.let { EventMode.valueOf(it) }
        val eventStatus = presenter.getEventStatus(eventUid)
        formView =
            FormView
                .Builder()
                .locationProvider(locationProvider)
                .onLoadingListener { loading: Boolean ->
                    if (loading) {
                        activity.showProgress()
                    } else {
                        activity.hideProgress()
                    }
                }.onItemChangeListener { action: RowAction ->
                    if (action.isEventDetailsRow) {
                        presenter.showOrHideSaveButton()
                    }
                    Unit
                }.onFinishDataEntry { presenter.saveAndExit(eventStatus) }
                .onFocused {
                    activity.hideNavigationBar()
                }.onPercentageUpdate { percentage: Float? ->
                    activity.updatePercentage(percentage!!)
                }.factory(activity.supportFragmentManager)
                .setRecords(EventRecords(eventUid, eventMode ?: EventMode.CHECK))
                .openErrorLocation(requireArguments().getBoolean(OPEN_ERROR_LOCATION, false))
                .setProgramUid(presenter.getEvent()?.program())
                .build()
        super.onCreate(savedInstanceState)
    }

    override fun onCreateView(
        inflater: LayoutInflater,
        container: ViewGroup?,
        savedInstanceState: Bundle?,
    ): View {
        binding =
            DataBindingUtil.inflate(inflater, R.layout.section_selector_fragment, container, false)
        val activityPresenter = activity.presenter
        binding.setPresenter(activityPresenter)

        activityPresenter.observeActions().observe(
            viewLifecycleOwner,
        ) { action: EventCaptureAction ->
            if (action == EventCaptureAction.ON_BACK) {
                formView?.onBackPressed()
                activityPresenter.emitAction(EventCaptureAction.NONE)
            }
        }

        binding.actionButton.setOnClickListener { view ->
            view.closeKeyboard()
            performSaveClick()
        }

        return binding.root
    }

    override fun onViewCreated(
        view: View,
        savedInstanceState: Bundle?,
    ) {
        super.onViewCreated(view, savedInstanceState)
        val transaction = childFragmentManager.beginTransaction()
        transaction.replace(R.id.formViewContainer, formView!!).commit()
        formView!!.scrollCallback = { isSectionVisible: Boolean ->
            animateFabButton(isSectionVisible)
        }
    }

    override fun onResume() {
        super.onResume()
        presenter.showOrHideSaveButton()
    }

    private fun animateFabButton(sectionIsVisible: Boolean) {
        var translationX = 1000
        if (sectionIsVisible) translationX = 0

        binding.actionButton
            .animate()
            .translationX(translationX.toFloat())
            .setDuration(500)
            .start()
    }

    override fun performSaveClick() {
        formView?.onSaveClick()
    }

    override fun hideSaveButton() {
        binding.actionButton.visibility = View.GONE
    }

    override fun showSaveButton() {
        binding.actionButton.visibility = View.VISIBLE
    }

    override fun onReopen() {
        formView?.reload()
    }

    override fun showNonEditableMessage(
        reason: String,
        canBeReOpened: Boolean,
    ) {
        binding.editableReasonContainer.visibility = View.VISIBLE

        binding.editableReasonContainer.setContent {
            NonEditableReasonBlock(
                modifier = Modifier.fillMaxWidth(),
                reason = reason,
                canBeReopened = canBeReOpened,
                onReopenClick = presenter::reOpenEvent,
            )
        }
    }

    override fun hideNonEditableMessage() {
        binding.editableReasonContainer.visibility = View.GONE
    }

    companion object {
        fun newInstance(
            eventUid: String?,
            openErrorSection: Boolean?,
            eventMode: EventMode,
        ): EventCaptureFormFragment {
            val fragment = EventCaptureFormFragment()
            val args = Bundle()
            args.putString(Constants.EVENT_UID, eventUid)
            openErrorSection?.let { args.putBoolean(OPEN_ERROR_LOCATION, it) }
            args.putString(Constants.EVENT_MODE, eventMode.name)
            fragment.arguments = args
            return fragment
        }
    }
}
